src/vdc_collection.rs

```diff
@@ -0,0 +1,205 @@
+pub enum CredentialFormat {
+    MDoc,
+    W3cJwtVc,
+    W3cJwtVcJsonld,
+    Other(String),  // For ease of expansion.
+}
+
+/// Supported credential types.
+#[derive(PartialEq)]
+pub enum CredentialType {
+    Iso18013_5_1mDl,
+    VehicleTitle,
+    Other(String),  // For ease of expansion.
+}
+
+/// An individual credential.  These are mainly for internal use; access to them should go through `VdcCollection`,
+/// below.
+struct Credential {
+    id:      String,           // Probably a UUID, perhaps we should make it one?
+    format:  CredentialFormat,
+    ctype:   CredentialType,
+    tags:    Vec<String>,      // Arbitrary string tags.
+    payload: Vec<u8>,          // The actual credential.
+    dirty:   bool,             // Has this been modified since it was read from storage?
+}
+
+impl Credential {
+    /// Create a new credential.
+    fn new(
+        id:      &str,
+        format:  CredentialFormat,
+        ctype:   CredentialType,
+        payload: Vec<u8>,
+        dirty:   bool,
+    ) -> Credential {
+        Credential {
+            id:     id.to_string(),
+            format,
+            ctype,
+            tags:   Vec::new(),
+            payload,
+            dirty,
+        }
+    }
+
+    /// Check whether a credential has a specific tag.
+    fn contains_tag(&self, tag: &str) -> bool {
+        for t in self.tags.iter() {
+            if t == tag {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Add a tag to a credential.
+    fn add_tag(&mut self, tag: &str) {
+        if !self.contains_tag(tag) {
+            self.tags.push(tag.to_string());
+            self.dirty = true;
+        }
+    }
+
+    /// Store a credential in secure storage if it has been modified.
+    fn write_to_secure_storage(&mut self) {
+        if self.dirty {
+            println!("Write {}", self.id);
+            // TODO: implement.
+
+            self.dirty = false;
+        }
+    }
+
+    /// Dump the contents of a credential.
+    fn dump(&self) {
+        println!("Credential {}", self.id);
+        print!("    format:   ");
+        match &self.format {
+            CredentialFormat::MDoc           => println!("MDoc"),
+            CredentialFormat::W3cJwtVc       => println!("W3C JWT VC"),
+            CredentialFormat::W3cJwtVcJsonld => println!("W3C JWT VC JSON LD"),
+            CredentialFormat::Other(x)       => println!("Other ({})", x),
+        }
+        print!("    type:     ");
+        match &self.ctype {
+            CredentialType::Iso18013_5_1mDl => println!("ISO 18013.5.1 mDL"),
+            CredentialType::VehicleTitle    => println!("Vehicle Title"),
+            CredentialType::Other(x)        => println!("Other ({})", x),
+        }
+        println!("    tags:");
+        for t in self.tags.clone() {
+            println!("        {t}");
+        }
+        if self.payload.is_empty() {
+            println!("    payload:  missing?");
+        }
+    }
```

Comment:
Instead of this you should `#[derive(Debug)]` on this struct and it will automatically generate equivalent code for displaying a debug view of the data.

Reply:
Done.